BOJ/2651.자동차경주대회/2651_자동차경주대회.cpp

```diff
@@ -0,0 +1,51 @@
+#include <iostream>
+#include <queue>
+using namespace std;
+
+
+
+int max_length;
+int end_stage;
+int distance_w[102];
+int cost[102];
+int memo[102];
+int memo_spot[102];
+//��� n���� ������������� ���ð�.
+int dp(int n) {
+	if (memo[n])
+		return memo[n];
+	if (n == end_stage)
+		return 0;
+
+	int res_min = 0;
+	int cnt_distance = 0;
+	for (int i = n + 1; i <= end_stage; i++) {
+		cnt_distance += distance_w[i];
+		if (cnt_distance > max_length)
+			break;
+		int temp = cost[n] + dp(i);
+		if (!res_min || res_min > temp) {
+			res_min = temp;
+			memo_spot[n] = i;
+		}
+	}
+	return memo[n] = res_min;
```

Comment:
```suggestion
	if (memo[n] || n == end_stage)
		return memo[n];



	int& res_min = memo[n];
	int cnt_distance = 0;
	for (int i = n + 1; i <= end_stage; i++) {
		cnt_distance += distance_w[i];
		if (cnt_distance > max_length)
			break;
		int temp = cost[n] + dp(i);
		if (!res_min || res_min > temp) {
			res_min = temp;
			memo_spot[n] = i;
		}
	}
	return res_min;
```
if문 조건은 하나로 합칠 수 있을 것 같아요. `res_min` 을 참조변수로 선언하면 `memo[n]` 에 대입하는 과정을 생략할 수 있습니다.

Reply:
memo가 전역이라 0으로 초기화되기때문에 가능한것인가요?
도움주셔서 감사합니다.

Reply:
@6047198844  가능하다는게 어떤걸 말씀하시는지 잘 모르겠어요..😥

Reply:
아 죄송합니다. 너무 요약해서 말씀드렸네요.
if (n == end_stage)
	return 0;
에서 memo[n]은 0으로 초기화된  전역변수이기 때문에 0을 반환하므로 0과 memo[n]은 동치. 따라서 다음의 코드는
if (n == end_stage)
	return memo[n];
과 같으므로
if (memo[n])
	return memo[n];
과 합칠수있음.
제가 질문했던건 memo[n]이 초기화된 전역변수라 0이기 때문에 두개의 if문을 합치는것이 가능한것인지 여쭤봤던거에요 . 제가 봐도 질문이 모호하네요.
답변 감사드립니다.

Reply:
@6047198844 아하 이해했습니다 ㅎㅎ
정확히는 `0` 과 `memo[end_stage]`가 동치입니다. `end_stage`인 상황에서 항상 0이 반환되며, `memo[end_stage]`의 값이 바뀔 일도 없습니다. 마침 `memo` 배열이 전역변수이니 0으로 초기화 되어 있으니 겸사겸사 두 개의 if문을 하나로 축약할 수 있게 됩니다. 그렇기때문에 합칠 수 있는 것이지요.